autoscheduler/frontend/src/redux/reducers/schedules.ts

```diff
@@ -32,69 +34,113 @@ export interface UnsaveScheduleAction {
+// where x is the lowest number schedule not already taken
+const createSchedule = (meetings: Meeting[], existingSchedules: Schedule[]): Schedule => ({
+  meetings,
+  name: getUniqueName(existingSchedules),
+  saved: false,
+});
+
+// Creates a copy of oldSchedules that will create new references for each schedule.
+// Note that the meetings array will still use the same reference.
+const cloneSchedules = (oldSchedules: Schedule[]): Schedule[] => (
+  [...oldSchedules].map((schedule) => ({ ...schedule }))
+);
+
 // returns whether a list of schedules contains a schedule with the same meetings as schedule
-export function containsSchedule(schedules: Meeting[][], schedule: Meeting[]): boolean {
+export function containsSchedule(allSchedules: Schedule[], schedule: Meeting[]): boolean {
   const scheduleMeetings = new Set(schedule.map((meeting) => meeting.id));
-  return schedules.some((toCompare) => {
-    const toCompareMeetings = new Set(toCompare.map((meeting) => meeting.id));
+  return allSchedules.some((toCompare) => {
+    const toCompareMeetings = new Set(toCompare.meetings.map((meeting) => meeting.id));
     return scheduleMeetings.size === toCompareMeetings.size
       && [...scheduleMeetings].every((meeting) => toCompareMeetings.has(meeting));
   });
 }
 
-function getUniqueSchedules(...schedules: Meeting[][]): Meeting[][] {
-  const unique: Meeting[][] = [];
-  schedules.forEach((schedule) => {
-    if (!containsSchedule(unique, schedule)) unique.push(schedule);
+function getUniqueSchedules(allSchedules: Schedule[]): Schedule[] {
+  const unique: Schedule[] = [];
+  allSchedules.forEach((schedule) => {
+    if (!containsSchedule(unique, schedule.meetings)) unique.push(schedule);
   });
   return unique;
 }
 
 // reducer
-function meetings(state: Schedules = initialSchedules, action: ScheduleAction): Schedules {
+function schedules(state: Schedule[] = initialSchedules, action: ScheduleAction): Schedule[] {
   switch (action.type) {
-    case ADD_SCHEDULE:
-      return { ...state, allSchedules: [...state.allSchedules, action.schedule] };
-    case REMOVE_SCHEDULE:
-      return {
-        ...state,
-        allSchedules: state.allSchedules.slice(0, action.index)
-          .concat(state.allSchedules.slice(action.index + 1)),
-      };
-    case REPLACE_SCHEDULES:
-      return {
-        ...state,
-        allSchedules: getUniqueSchedules(...state.savedSchedules, ...action.schedules),
-      };
-    case SAVE_SCHEDULE:
-      return {
-        ...state,
-        savedSchedules: state.savedSchedules
-          .concat(containsSchedule(state.savedSchedules, state.allSchedules[action.index])
-            ? [] : [state.allSchedules[action.index]]),
-      };
-    case UNSAVE_SCHEDULE:
-      return {
-        ...state,
-        savedSchedules: state.savedSchedules.filter((schedule) => (
-          schedule !== state.allSchedules[action.index])),
-      };
+    case ADD_SCHEDULE: {
+      const newState = cloneSchedules(state);
```

Comment:
I had read in a few places that it's good practice with redux to deep copy state, so I changed it to do that. Then I read this comment and found a few other articles say that you should only deep copy attributes that are changed (which makes a lot of sense, as it avoids performance degradation from all the copying and still supports time travel as long as you implement it correctly). I think changing it back to not clone and just do a shallow copy would be better, what do you think?

Reply:
I agree that shallow copy should be fine if we're not actually changing anything but it seems like you are changing the state here? When you push a new schedule onto it, we might need a new object reference to let Redux/React know that we're updating the schedules state

Reply:
> 
> 
> I agree that shallow copy should be fine if we're not actually changing anything but it seems like you are changing the state here? When you push a new schedule onto it, we might need a new object reference to let Redux/React know that we're updating the schedules state

Yeah we do need new references when for example the name is changed for a particular schedule, but we don't need to make new references for every schedule all the time. Here's all the redux changes I made to switch it to shallow copying:
![image](https://user-images.githubusercontent.com/28655462/94346489-967a7e00-fff2-11ea-9dd2-fe4878a1c6f9.png)
This is a bit faster for the reasons in my first comment, so I can push a commit with those changes if we want

Reply:
Not sure what the `createSchedule` function does, but it probably doesn't matter tbh. Based on what I can see in this picture, your code seems equivalent to the best practices outlined on [the Redux site](https://redux.js.org/recipes/structuring-reducers/immutable-update-patterns), so I'd say we go back to shallow copy

Reply:
> Not sure what the `createSchedule` function does, but it probably doesn't matter tbh

Yeah it's not relevant to this, but since schedules used to be passed in as meeting arrays and the schedules state was a `Meeting[][]` I decided to keep the interface for actions the same and have `createSchedule` make an object of schedule type that takes in the meetings and initializes all the other needed props, like `name` and `saved`